crates/bevy_ecs/macros/src/lib.rs

```diff
@@ -248,16 +247,29 @@ pub fn impl_param_set(_input: TokenStream) -> TokenStream {
     tokens
 }
 
-#[derive(Default)]
-struct SystemParamFieldAttributes {
-    pub ignore: bool,
+/// Checks if the specified identifier occurs in the specified [`TokenStream`].
+#[allow(clippy::cmp_owned)]
+fn check_for_collision(haystack: TokenStream, value: &Ident) -> bool {
+    haystack.into_iter().any(|t| match t {
+        TokenTree::Group(g) => check_for_collision(g.stream(), value),
+        TokenTree::Ident(ident) => ident.to_string() == value.to_string(),
+        TokenTree::Punct(_) | TokenTree::Literal(_) => false,
+    })
 }
 
-static SYSTEM_PARAM_ATTRIBUTE_NAME: &str = "system_param";
+fn ensure_no_collision(haystack: TokenStream, mut value: Ident) -> Ident {
+    // If there's a collision, add more characters to the identifier
+    // until it doesn't collide with anything anymore.
+    while check_for_collision(haystack.clone(), &value) {
```

Comment:
This potentially iterates through the TokenStream an arbitrary number of times for each call to this function, and may be done multiple times as `ensure_no_collision` may be used multiple times across a single macro. Can we choose a data structure here that has better reuse characteristics?